onnxruntime/core/session/inference_session.cc

```diff
@@ -927,6 +873,13 @@ common::Status InferenceSession::Initialize() {
     // now that all the transforms are done, call Resolve on the main graph. this will recurse into the subgraphs.
     ORT_RETURN_IF_ERROR_SESSIONID_(graph.Resolve());
 
+    // need to keep the initializers if we're going to save the optimized model
+    bool keep_initializers = !session_options_.optimized_model_filepath.empty();
+
```

Comment:
This change makes it cleaner to support saving an updated model in a different format that needs to be done after finalization (vs. old code of saving and then finalizing). 

Assumption: If you're saving the optimized model it's not a runtime scenario, so leaving the initializers in the model after they're copied to SessionState is of no consequence.